rust/kernel/fwnode.rs

```diff
@@ -0,0 +1,111 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Unified device property interface.
+//!
+//! C header: [`include/linux/property.h`](../../../../include/linux/property.h)
+
+use crate::{bindings, device, sync::Arc};
+use core::ptr::null_mut;
+
+fn next_endpoint(handle: &DeviceNode, curr: &Option<Arc<Node>>) -> Option<Arc<Node>> {
+    let res_ptr = match curr {
+        // SAFETY: By the type invariants, `handle` owns a reference, so it is safe to access `ptr`
+        None => unsafe {
+            bindings::fwnode_graph_get_next_endpoint(
+                handle.ptr,
+                null_mut::<bindings::fwnode_handle>(),
+            )
+        },
+        // SAFETY: By the type invariants, `handle` owns a reference, so it safe to access it's
+        // `ptr`. `node` is a reference counted object (Arc) and hence its safe to access the
+        // internal `ptr`
+        Some(node) => unsafe {
+            bindings::fwnode_graph_get_next_endpoint(handle.ptr, node.as_ref().ptr)
+        },
+    };
+    if res_ptr.is_null() {
+        None
+    } else {
+        Arc::try_new(Node { ptr: res_ptr }).ok()
+    }
+}
+
+/// Represents the root `struct fwnode_handle *` associated to a device
+///
+/// # Invariants
+///
+/// The pointer is valid.
+pub struct DeviceNode {
+    ptr: *mut bindings::fwnode_handle,
+}
+
+impl DeviceNode {
+    /// Creates a new DeviceNode instance from an existing [`device::Device`] instance.
+    pub fn from(dev: &device::Device) -> Self {
+        Self {
+            // SAFETY: By the type invariants, `dev` owns a reference, so it is safe to access
+            // `ptr`
+            ptr: unsafe { bindings::dev_fwnode(dev.ptr) },
+        }
+    }
+    /// Creates an instance of `NodeIterator`
+    ///
+    /// This provides an `Iterator` wrapping the internal C functionality of invoking
+    /// `fwnode_graph_get_next_endpoint`
+    pub fn endpoints(self) -> NodeIterator {
+        NodeIterator {
+            handle: self,
+            curr_node: None,
+            iter_fn: next_endpoint,
+        }
+    }
+}
+
+// SAFETY: `DeviceNode` only holds a pointer to a C fwnode_handle, which is safe to be used from any thread.
+unsafe impl Send for DeviceNode {}
+
+// SAFETY: `DeviceNode` only holds a pointer to a C fwnode_handle, references to which are safe to be used
+// from any thread.
+unsafe impl Sync for DeviceNode {}
+impl Drop for DeviceNode {
+    fn drop(&mut self) {
+        // SAFETY: By the type invariants, `self` owns a reference, so it is safe to
+        // relinquish it now.
+        unsafe { bindings::fwnode_handle_put(self.ptr) }
+    }
+}
+
+/// Represents a `struct fwnode_handle *` part of a device's fwnode graph
+///
+/// # Invariants
+///
+/// The pointer is valid.
+pub struct Node {
+    pub(crate) ptr: *mut bindings::fwnode_handle,
```

Comment:
I think this should store `UnsafeCell<bindings::fwnode_handle>` and be `#[repr(transparent)]` directly and implement `AlwaysRefCounted`. Then you can return an `ARef<Node>` from `NodeIterator`. See for example https://github.com/Rust-for-Linux/linux/blob/729c03011c60382030520f84904a85e8a5b3cfb9/rust/kernel/file.rs#L113-L179

Reply:
thanks @bjorn3 for the review! Is it correct to reason that this is not required for DeviceNode since it's ptr is not public? And hence its safe to rely on rust's reference counting completely.

Reply:
Also I don't know if you endpoints of nodes can have endpoints themself, but if so maybe `DeviceNode` should be merged into `Node`.

Reply:
> if you endpoints of nodes can have endpoints themself,

I took your recommendation and merged `DeviceNode` and `Node`. Based on the [documentation](https://www.kernel.org/doc/html/latest/firmware-guide/acpi/dsd/graph.html) endpoints should not have endpoints themselves. However, calling `endpoints()` returns an Iterator which may be empty as well, so should be safe for consumption. 

> dev_fwnode seems to return a pointer without incrementing reference count.

I have added a `fwnode_handle_get` call to ensure that the refcount is incremented before passing it to the ARef. Since I am explicitly incrementing the refcount in `Node::from` I am not associating the lifetime of `Node` with the device. Am i missing anything?

Reply:
> > dev_fwnode seems to return a pointer without incrementing reference count.
> 
> I have added a `fwnode_handle_get` call to ensure that the refcount is incremented before passing it to the ARef. Since I am explicitly incrementing the refcount in `Node::from` I am not associating the lifetime of `Node` with the device. Am i missing anything?

Should be fine now.